### PR TITLE
Implement getPublicTimetables endpoint

### PR DESCRIPTION
**Pull request by palaniappan-r**
branch[] , year and sem should be passed as query parameters. If query parameters are not passed, all timetables will be returned.

### CODE REVIEW

File: src/controllers/timetable/getPublicTimetables.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        if(sem){
+            try {
+                queryBuilder.andWhere("timetable.semester = :sem" , { sem })
+            } catch (err: any) {
+                console.log("Error while querying timetable: ", err.message);
+                res.status(500).json({ message: "Internal Server Error" });
+            }
+        }
+
+        timetables = await queryBuilder.getMany();
```
Comment (majimearun) [inline src/controllers/timetable/getPublicTimetables.ts:95]: You don't have to wrap each where clause in a try-ctach block. The error message should contain the required information regardless. Wrap each db interaction/query in one try-ctach. If there is something you read which told you that this is the right practice, do share. So the rest of us can follow it as well.

File: src/controllers/timetable/getPublicTimetables.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import { timetableRepository } from "../../repositories/timetableRepository";
+import { z } from "zod";
+import { validate } from "../../utils/zodValidateRequest";
+
+import {
+    isAValidDegreeCombination,
+  } from "../../types/degrees";
+
+const dataSchema = z.object({
+    query : z.object({
```
Comment (majimearun) [inline src/controllers/timetable/getPublicTimetables.ts:12]: No auth temp replacement or using the cookie either?

File: src/controllers/timetable/getPublicTimetables.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                invalid_type_error : "sem is not a number"
+            })
+            .gte(1 , {
+                message : "invalid sem number (can only be 1 or 2)"
+            })
+            .lte(2 , {
+                message : "invalid sem number (can only be 1 or 2)"
+            })
+            .optional(),
+        branch: z.array(z.string())
```
Comment (majimearun) [inline src/controllers/timetable/getPublicTimetables.ts:35]: We already have a degree enum, use that to validate

File: src/controllers/timetable/createTimetable.ts [hunk 4]
```diff
@@ -80,7 +80,7 @@ export const createTimetable = async (req: Request, res: Response) => {
           examEndTimes,
           warnings,
           createdAt,
-          lastUpdated,
+          lastUpdated
```
Comment (majimearun) [inline src/controllers/timetable/createTimetable.ts:83]: The `,` is from prettier which is we're using for uniformity lol

**Review by majimearun — CHANGES_REQUESTED** [review]
Besides the small changes requested, looks good. Like mentioned in the comments, if there's a reason why you've written queries the way you did, please mention it

File: src/controllers/timetable/getPublicTimetables.ts (hunk 1 above)
Comment (palaniappan-r) [reply src/controllers/timetable/getPublicTimetables.ts]: You're right. Those try/catch blocks are redundant. Removed them now.

File: src/controllers/timetable/getPublicTimetables.ts (hunk 3 above)
Comment (palaniappan-r) [reply src/controllers/timetable/getPublicTimetables.ts]: Done

**Comment by soumitradev** [comment]
I'll be reviewing this, since aruna is busy on #28

File: src/controllers/timetable/getPublicTimetables.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    isAValidDegreeCombination,
+  } from "../../types/degrees";
+
+const dataSchema = z.object({
+
+    query : z.object({
+        // temp auth replacement
+        email: z.string({
+            invalid_type_error: "email not a string",
+            required_error: "email is a required path parameter",
```
Comment (soumitradev) [inline src/controllers/timetable/getPublicTimetables.ts:20]: check #30 for a more fleshed-our version of these errors

File: src/controllers/timetable/getPublicTimetables.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            }).max(2, {
+            message: "branch may not contain more than two elements",
+            }),
+
+    })
+})
+
+export const getPublicTimetablesValidator = validate(dataSchema)
+
+export const getPublicTimetables = async (req : Request , res : Response) => {
```
Comment (soumitradev) [inline src/controllers/timetable/getPublicTimetables.ts:55]: I don't think your code is formatted correctly, use Prettier

File: src/controllers/timetable/getPublicTimetables.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            .select(["timetable.id", "timetable.name", "timetable.lastUpdated", "timetable.createdAt", "timetable.degrees"])
+            .where("timetable.private = :isPrivate" , { isPrivate })
+
+        if(branch){
+            if (branch.length === 2 && !isAValidDegreeCombination(branch)) {
+                return res.status(400).json({
+                message: "Branch may only have one valid BE degree and one valid MSc degee",
+                });
+            }
+            queryBuilder.andWhere("timetable.degrees = :branch", { branch })
```
Comment (soumitradev) [inline src/controllers/timetable/getPublicTimetables.ts:91]: I'm pretty sure you have to reassign queryBuilder to the return value of the function, you can't just call the function.

File: src/controllers/timetable/getPublicTimetables.ts [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+          }
+
+        let branch: DegreeList = req.query.branch as DegreeList;
+        let year: number = parseInt(req.query.year as string);
+        let sem: number = parseInt(req.query.sem as string);
+        let isPrivate: boolean = false
+
+        const queryBuilder = timetableRepository
+            .createQueryBuilder("timetable")
+            .select(["timetable.id", "timetable.name", "timetable.lastUpdated", "timetable.createdAt", "timetable.degrees"])
```
Comment (soumitradev) [inline src/controllers/timetable/getPublicTimetables.ts:82]: check #30 for the exact fields you need to select for timetable metadata

**Review by soumitradev — CHANGES_REQUESTED** [review]
format your code using Prettier, and resolve some other issues

File: src/controllers/timetable/getPublicTimetables.ts (hunk 8 above)
Comment (palaniappan-r) [reply src/controllers/timetable/getPublicTimetables.ts]: I'm selecting these fields for now, is this alright?

```
"timetable.id",
"timetable.name",
"timetable.degrees",
"timetable.private",
"timetable.draft",
"timetable.archived",
"timetable.acadYear",
"timetable.year",
"timetable.semester",
"timetable.createdAt",
"timetable.lastUpdated"
```

File: src/controllers/timetable/getPublicTimetables.ts (hunk 7 above)
Comment (palaniappan-r) [reply src/controllers/timetable/getPublicTimetables.ts]: It worked fine with timetables created by the `createTimetable` endpoint. [I manually added a dummy user in the db]

I've reassigned now though.

File: src/controllers/timetable/getPublicTimetables.ts (hunk 8 above)
Comment (soumitradev) [reply src/controllers/timetable/getPublicTimetables.ts]: yes

**Review by soumitradev — DISMISSED** [review]
LGTM

**Comment by soumitradev** [comment]
resolving merge conflict, and merging

File: src/controllers/timetable/createTimetable.ts
Comment (soumitradev) [inline src/controllers/timetable/createTimetable.ts:1]: Actually can you revert the change to this file? I'll merge after that

**Review by soumitradev — CHANGES_REQUESTED** [review]

**Review by soumitradev — APPROVED** [review]

File: src/controllers/timetable/createTimetable.ts
Comment (soumitradev) [reply src/controllers/timetable/createTimetable.ts:1]: I formatted it instead

**Pull request merged**